### PR TITLE
Add results schema validation

### PR DESCRIPTION
**Pull request by StevePole**

### CODE REVIEW

File: src/api/routes/test.test.ts [hunk 1]
```diff
@@ -23,28 +27,113 @@ describe('test endpoints', () => {
...
+          testTypeId: testType.id.value,
+          results: {
+            testerUserId: user.id.value,
+            details: { c: 12, igg: 'Wot', igm: [] },
+          },
+        })
+        .expect(422);
+    });
+
+    it('returns 200 with the existing test if user is found', async () => {
```
Comment (kstam) [inline src/api/routes/test.test.ts]: Couple of comments on this:
1. I would expect POST /resource to return 201 Created on success
2. Given this is a new test the test name should say something along the lines of  `returns 201 with the newly created test if user is found`

File: src/api/routes/test.test.ts [hunk 2]
```diff
@@ -23,28 +27,113 @@ describe('test endpoints', () => {
...
+      const user = await userRepository.save(aNewUser());
+      const unknowUserId = new UserId();
+
+      await request(app)
+        .post(`/api/v1/users/${unknowUserId.value}/tests`)
+        .set({ Authorization: `Bearer ${await getTokenForUser(user)}` })
+        .expect(404);
+    });
+
+    it('returns 403 if a user with an access pass tries to create a test for another user', async () => {
```
Comment (kstam) [inline src/api/routes/test.test.ts]: Should we not allow this? If I am a pharmasist that is entering a new test for someone shouldn't I be able to create a test for them?

File: src/application/service/CreateOrUpdateTest.ts [hunk 3]
```diff
@@ -1,33 +1,53 @@
 import { UserId } from '../../domain/model/user/UserId';
 import { Test } from '../../domain/model/test/Test';
 import { TestId } from '../../domain/model/test/TestId';
+import { Results } from '../../domain/model/test/Results';
 import { TestRepository } from '../../domain/model/test/TestRepository';
 import { getTestTypes } from './index';
 import { TestTypeId } from '../../domain/model/testType/TestTypeId';
-import { Test as ApiTest } from '../../api/interface/index';
+import { TestCommand } from '../../api/interface/index';
+import jsonschema from 'jsonschema';
```
Comment (kstam) [inline src/application/service/CreateOrUpdateTest.ts]: Given we already have AJV which we need to validate the meta-schema couldn't we also use it to validate this schema? From what I saw 'jsonschema' doesn't allow meta-schema validation

File: src/api/routes/test.test.ts (hunk 2 above)
Comment (StevePole) [reply src/api/routes/test.test.ts]: Yeah, typo.  It's fixed in the follow up PR that actually implements it.

Comment (kstam) [reply src/api/routes/test.test.ts]: Should we fix it here so we can merge to master and move on.. Doesn't matter that much given it's coming right after. up to you

File: src/infrastructure/persistence/PsqlTestRepository.ts [hunk 4]
```diff
@@ -50,6 +50,7 @@ export class PsqlTestRepository implements TestRepository {
       new TestId(linkRow.id),
       new UserId(linkRow.userId),
       new TestTypeId(linkRow.testTypeId),
+      undefined,
```
Comment (StevePole) [inline src/infrastructure/persistence/PsqlTestRepository.ts]: Note, this looks weird!  At this point, the persistence for results is not implemented (comes in follow on PR).  I rearranged the order of the constructor params, so needed this temporarily.

File: src/application/service/CreateOrUpdateTest.ts (hunk 3 above)
Comment (StevePole) [reply src/application/service/CreateOrUpdateTest.ts]: I didn't realise we had something that could do it.  Harsh was importing this one in package.json already.  Although I think he was only doing it with this use case in mind.

File: src/infrastructure/persistence/PsqlTestRepository.ts (hunk 4 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlTestRepository.ts]: Yep I saw this in the following PR

File: src/application/service/CreateOrUpdateTest.ts (hunk 3 above)
Comment (StevePole) [reply src/application/service/CreateOrUpdateTest.ts]: I'll switch.  Simpler API as well.

File: src/api/routes/test.test.ts (hunk 2 above)
Comment (StevePole) [reply src/api/routes/test.test.ts]: Then I'll have to deal with a conflict so easier just to proceed!

File: src/application/service/CreateOrUpdateTest.ts [hunk 5]
```diff
@@ -1,33 +1,58 @@
...
     const { testTypeId, results } = payload;
-    const checkTestType = await getTestTypes.byId(testTypeId);
+    const testType = await getTestTypes.byId(testTypeId);
+
+    if (!testType) {
+      throw new ResourceNotFoundError('testTypeId', testTypeId);
+    }
+
+    if (results) {
+      const isValid = jsonSchema.validate(
```
Comment (kstam) [inline src/application/service/CreateOrUpdateTest.ts]: Sorry I missed this before  but wouldn't it be better to move this validation to the domain? 

I also talked briefly about this with Harsh at the time. 

We can introduce a TestResultsFactory which can be instantiated with a testTypeRepository. 

Then TestResultFactory can have a TestFactory#createTestResults which only ever creates valid Results objects. 

This way this code can be moved to the application and domain package and we will guarantee that there is no way to create an invalid TestResult

```

```

**Pull request merged**